Cards in dashboards: milestone 1 (backend)

johnswanson opened the pull request:
- migrations to add `report_card.dashboard_id`, and an index/foreign key for it
- accept a `dashboard_id` parameter for `POST /api/card/`
- accept a `dashboard_id` parameter for `PUT /api/card/`
- return `dashboard_id` from the search API
- do not show dashboard questions without a DashboardCard in the search API
- do not show dashboard questions in the autocomplete API
- modifies `POST /api/dashboard/:id/copy` to always copy dashboard questions
- modifies `PUT /api/dashboard/:id` to automatically move dashboard questions along with the dashboard
- modifies the Search API to allow including (or not) dashboard questions

Product doc: https://www.notion.so/metabase/Enable-saving-questions-directly-to-dashboards-and-make-this-the-default-path-a84be6e0540241189b251401fb35ca91#f9bbd61d8bc442ab8eaf908c680394b4

Epic: https://github.com/metabase/metabase/issues/47326

johnswanson committed:
Migration to add `report_card.dashboard_id`

Setting this makes the card a "dashboard internal card" that should only
be used within a single dashboard, is moved along with the dashboard, etc.

johnswanson committed:
prevent invalid inserts/updates

johnswanson committed:
POST /api/card for dashboard internal cards

johnswanson committed:
GET /api/collection/:id/items

johnswanson committed:
Search

johnswanson committed:
PUT /api/card/:id

johnswanson committed:
WIP: last bits of card-in-dashboard API

Sticking point here is the copy, I think I need to do a small refactor
to make the code less hairy.

johnswanson committed:
rm line

johnswanson committed:
optionally show DQs for collection items

johnswanson committed:
copying dashboards :tada:

johnswanson committed:
wip

crisptrutski commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -9270,6 +9270,90 @@ databaseChangeLog:
 
+
+  - changeSet:
+      id: v51.2024-08-28T07:03:41
```
Following the discussion in our last guild, should we not add a descriptive suffix to these ids?

johnswanson replied:
I don't think we've updated the linter with that ability yet?

crisptrutski commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -9270,6 +9270,90 @@ databaseChangeLog:
+                  name: is_dashboard_internal_card
+                  type: ${boolean.type}
+                  remarks: |
+                    A denormalized representation that this edge from dashboard<->card should be the *only* edge from its card.
+                    Provides an efficient way to enforce the fact that
```
Watch out for this trailing sentence. How about the following?

```suggestion
                    A denormalized edge from dashboard<->card which uses a unique index to ensure that
                    there is at most one `is_dashboard_internal_card` dashcard per card.
```

This also seems to be attached to the wrong changeset, and belongs on the next one.

johnswanson replied:
I ended up deciding against creating this column at all, as it was confusing and painful to keep up-to-date and didn't provide a lot of value (e.g. as you pointed out, if we move the dashboard card to a new dashboard or a collection, we'd need to update the dashcards as well).

crisptrutski commented on resources/migrations/001_update_migrations.yaml:
```diff
@@ -9270,6 +9270,90 @@ databaseChangeLog:
+  - changeSet:
+      id: v51.2024-08-28T07:03:43
+      author: johnswanson
+      comment: Add `report_dashboardcard.dashboard_internal_card_id`
```
This unique field won't guarantee that there aren't some non-`is_dashboard_internal_card` dashcards for the given card 🤔

crisptrutski replied:
Ah, if I understand the product right, this is not an issue?

If we move a "dashboard card" to a collection, would we leave the `dashboard_internal` fields in place?

johnswanson replied:
See above, I [ended up removing this column entirely](https://github.com/metabase/metabase/pull/47369#discussion_r1759449149).

crisptrutski reviewed (commented):
~~Can you drop a link to the epic or project doc in the description? I'm curious to understand more about the internal card feature.~~ https://www.notion.so/metabase/Cards-in-Dashboards-39c889612c094ff2be3c7c8e291a738b

crisptrutski commented:
PS sorry if I'm reviewing prematurely, I see it was automation on the migrations file that had added BEC to review.

johnswanson committed:
quick fix

johnswanson committed:
Disallow moving dashboard questions & test archive

johnswanson committed:
remove unused columns

johnswanson committed:
fix argument order :facepalm:

johnswanson committed:
serdes for dashboard questions

johnswanson committed:
serialization baseline tests

johnswanson committed:
Fix tests in `metabase.models.card`

johnswanson committed:
Fix download exports test

johnswanson committed:
Lint migrations file

johnswanson committed:
cljfmt

johnswanson committed:
cljfmt, again

the cljfmt job in CI doesn't run on this branch, it runs on the PR
branch merged with `origin/master` - which is nice, because it catches
cljfmt issues that would happen if you merged, but don't happen on your
branch, but also confusing, because it catches issues like
"metabase.api.common is never used", and then you stare at the file for
a while like "YES IT IS" before realizing what's going on.

johnswanson committed:
migration to add idx for report_card.dashboard_id

johnswanson committed:
don't rollback the `report_card.dashboard_id` bits

johnswanson committed:
a couple missed API changes

- no DQs w/o DashboardCards in search

- no DQs in autocomplete

- fix check for "currently a dashboard question" in card validation

johnswanson committed:
Fix some comments, clean up some existing code

Before, `cards-to-copy` returned just two lists of cards: those to
discard (no permissions) and those to copy. But `maybe-duplicate-cards!`
had some extra logic: if the card was a `model`, don't copy it, just
reference it.

I modified `cards-to-copy` to return three lists instead: cards to copy,
cards to discard, *and* cards to reference. This was initially just used
for making sure we copy Dashboard Questions instead of referencing them
- but I realized we could use the same for Models as well. I moved the
logic for models to `cards-to-copy` instead: if the card is a Model,
return it in the `reference` list instead of the `copy` list.

johnswanson committed:
Fix commented-out test

johnswanson committed:
cljfmt

johnswanson committed:
Optionally include dashboard questions in search

johnswanson committed:
Add a test for failing card name change

johnswanson committed:
Opt. include dashboard questions for autocomplete

noahmoss commented on src/metabase/api/dashboard.clj:
```diff
@@ -611,9 +622,18 @@
     (dashboard-card/delete-dashboard-cards! dashcard-ids)
     dashboard-cards))
 
+(defn- assert-no-invalid-dashboard-internal-dashcards [dashboard to-create]
```
This function name took me a solid minute to understand. Maybe `assert-new-dashcards-are-not-internal-to-other-dashboards`? Or just add a comment

noahmoss replied:
I think the problem is that "dashboard-internal dashcards" make sense as a phrase, but the grouping is lost when the whole thing is kebab case

noahmoss commented on src/metabase/models/card.clj:
```diff
@@ -400,11 +400,62 @@
+
+(def ^:dynamic ^:private *updating-dashboard* false)
+
+(defmacro with-allowed-changes-to-dashboard-internal-card
```
This (and other functions in this namespace) have a similar issue to what I mentioned above. The phrase "dashboard-internal" is hard to make sense of when it's in the middle of a larger kebab-case function name. It's fine when you're already familiar with the concept but hard to parse with fresh eyes.

Not sure what the best solution is. Maybe just flip it around as `internal dashboard card`?

johnswanson replied:
I think flipping it around makes sense. It's still not totally clear, but seems a bit clearer, so seems good for now. Maybe in the long run we can try to figure out a better name for "Dashboard Questions", "Internal Dashboard Cards", "Dashboard Internal Cards", etc. Feels like right now we have lots of different names, all of which are a bit confusing and some of which have possibly overloaded meanings (e.g. "Dashboard Questions" feels right to me, but seems easy to confuse with DashboardCards).

noahmoss commented on src/metabase/search/impl.clj:
```diff
@@ -232,6 +232,13 @@
                               [:= :bookmark.user_id (:current-user-id search-ctx)]])
+      (sql.helpers/where [:or
+                          [:= nil :card.dashboard_id]
+                          (when (:include-dashboard-questions? search-ctx)
+                            [:exists
+                             {:select 1
+                              :from [:report_dashboardcard]
+                              :where [:= :card_id :card.id]}])])
```
I don't understand the logic here. Why not just omit the `:= nil :card.dashboard_id` check when we want to include dashboard questions?

(If there's a reason, could you make it clear in the code?)

johnswanson replied:
Good call - I added some explanatory comments. The oddity here is that if we end up with a Dashboard Question that has no DashboardCards, we want it to "disappear" from the instance. This is kind of an artifact of the weirdness of having a Dashboard Question be a separate entity from the actual link to the Dashboard itself - in an ideal world, it would be impossible to represent the situation where you have a DQ but it's not present on a dashboard. But... because it is possible to represent that situation (and it might happen, e.g. if something goes wrong during the card creation process), we want to handle it correctly and hide it from search.

noahmoss commented on test/metabase/api/card_test.clj:
```diff
@@ -3847,3 +3847,56 @@
+      (mt/user-http-request :crowberto :post 400 "card" (assoc (card-with-name-and-query)
+                                                               :dashboard_id dash-id
+                                                               :collection_id other-coll-id)))
+    (testing "Note the bug: if you provide `collection_id=null`, we'll accept it."
```
Wdym by bug? Is this something that should be changed, or expected behavior?

johnswanson replied:
Ah, good callout. I had meant to come back to this.

Normally if you provide a `collection_id` we require that it's the same as the `collection_id` of the dashboard you're telling us to put the card in, and respond with a 4XX otherwise.

But in the case where you provide `collection_id=null`, and the dashboard _isn't_ in the root collection, we ignore the mismatch. We don't _do_ anything crazy - we just put the card into the dashboard's collection, not the root collection, and ignore the `collection_id` you passed in (it's effectively overridden by the `dashboard_id`).

The reason for this is that by the time we're in the before-insert toucan hook, we've set any missing values to `nil` in `metabase.models.card/create-card!` - but I've just looked it over and I think I can avoid that pretty easily. We were just doing that for the purpose of calling `api/maybe-reconcile-collection-position!`, which needs `collection_id` and `collection_position` to be present. I'll just rework things to keep the card-data the same (with missing keys) for `t2/insert!` and call `api/maybe-reconcile-collection-position!` with a different map. Then this bug will be gone and things should be consistent.

noahmoss reviewed (approved):
Awesome! Just some nits/questions/semantic comments

johnswanson committed:
Update src/metabase/api/dashboard.clj

Co-authored-by: Noah Moss <32746338+noahmoss@users.noreply.github.com>

johnswanson committed:
PR review response

johnswanson committed:
lint migrations
